tests/jsunit/workspace_undo_redo_test.js

```diff
@@ -0,0 +1,386 @@
+  workspace.createVariable('name1', 'type1', 'id1');
+  workspace.createVariable('name2', 'type2', 'id2');
+  workspace.deleteVariableById('id1');
+  workspace.deleteVariableById('id2');
+
+  workspace.undo();
+  assertNull(workspace.getVariableById('id1'));
+  undoRedoTest_checkVariableValues('name2', 'type2', 'id2');
+
+  workspace.undo();
+  undoRedoTest_checkVariableValues('name1', 'type1', 'id1');
+  undoRedoTest_checkVariableValues('name2', 'type2', 'id2');
+  undoRedoTest_tearDown();
+}
+
+function test_undoDeleteVariable_WithBlocks() {
+  undoRedoTest_setUpWithMockBlocks();
+  workspace.createVariable('name1', 'type1', 'id1');
+  workspace.createVariable('name2', 'type2', 'id2');
+  createMockBlock('name1');
+  createMockBlock('name2');
+  workspace.deleteVariableById('id1');
+  workspace.deleteVariableById('id2');
+
+  workspace.undo();
+  undoRedoTest_checkBlockVariableName(0, 'name2');
+  assertNull(workspace.getVariableById('id1'));
+  undoRedoTest_checkVariableValues('name2', 'type2', 'id2');
+
+  workspace.undo();
+  undoRedoTest_checkBlockVariableName(0, 'name2');
+  undoRedoTest_checkBlockVariableName(1, 'name1');
+  undoRedoTest_checkVariableValues('name1', 'type1', 'id1');
+  undoRedoTest_checkVariableValues('name2', 'type2', 'id2');
+  undoRedoTest_tearDownWithMockBlocks();
+}
+
+function test_redoAndUndoDeleteVariable() {
+  undoRedoTest_setUp();
+  workspace.createVariable('name1', 'type1', 'id1');
+  workspace.createVariable('name2', 'type2', 'id2');
+  workspace.deleteVariableById('id1');
+  workspace.deleteVariableById('id2');
+
+  workspace.undo();
+  workspace.undo(true);
+  // Expect that both variables are deleted
+  assertNull(workspace.getVariableById('id1'));
+  assertNull(workspace.getVariableById('id2'));
+
+  workspace.undo();
+  workspace.undo();
+  workspace.undo(true);
+  // Expect that variable 'id2' is recreated
+  assertNull(workspace.getVariableById('id1'));
+  undoRedoTest_checkVariableValues('name2', 'type2', 'id2');
+  undoRedoTest_tearDown();
+}
+
+function test_redoAndUndoDeleteVariableWithBlocks() {
+  undoRedoTest_setUpWithMockBlocks();
+  workspace.createVariable('name1', 'type1', 'id1');
+  workspace.createVariable('name2', 'type2', 'id2');
+  createMockBlock('name1');
+  createMockBlock('name2');
+  workspace.deleteVariableById('id1');
+  workspace.deleteVariableById('id2');
+
+  workspace.undo();
+  workspace.undo(true);
+  // Expect that both variables are deleted
+  assertEquals(0, workspace.topBlocks_.length);
+  assertNull(workspace.getVariableById('id1'));
+  assertNull(workspace.getVariableById('id2'));
+
+  workspace.undo();
+  workspace.undo();
+  workspace.undo(true);
+  // Expect that variable 'id2' is recreated
+  undoRedoTest_checkBlockVariableName(0, 'name2');
+  assertNull(workspace.getVariableById('id1'));
+  undoRedoTest_checkVariableValues('name2', 'type2', 'id2');
+  undoRedoTest_tearDownWithMockBlocks();
+}
+
+function test_undoRedoRenameVariable_NeitherVariableExists() {
```

Comment:
Added an explaining comment. For reference it is:
"Expect that a variable with the name, 'name2', and the generated UUID, 'id2', to be created when rename is called. Undo removes this variable and redo recreates it."